Code/GraphMol/Depictor/EmbeddedFrag.cpp

```diff
@@ -346,14 +346,151 @@ void EmbeddedFrag::setupAttachmentPoints() {
+  // it crosses an odd number of times, the point is inside the ring
+  float outsideX = minPt.x - 0.1;
+  int intersections = 0;
+  for (size_t i = 0; i < ring.size(); ++i) {
+    RDGeom::Point2D p1 = coords.at(ring[(i)]);
+    RDGeom::Point2D p2 = coords.at(ring[(i + 1) % ring.size()]);
+    if (crossing(p1, p2, pt, RDGeom::Point2D(outsideX, pt.y))) {
+      ++intersections;
+    }
+  }
+  return (intersections % 2) == 1;
+}
+
+static float scoreMatch(const RDKit::VECT_INT_VECT &fusedRings,
+                        const RDKit::INT_VECT &ringSystemAtoms,
+                        const RDGeom::INT_POINT2D_MAP &ringSystemAtomsCoords,
+                        const RDKit::ROMol *mol) {
+  // for each atom in rs_mol that has neighbors that are not in the match, add a
+  // penalty if the corresponding atom in the template points inside a ring.
+  int result = 0;
+  for (auto index : ringSystemAtoms) {
+    auto neighbors = mol->atomNeighbors(mol->getAtomWithIdx(index));
+    int neighborsOutsideTemplate = 0;
+    RDKit::INT_VECT neighborsInsideTemplate;
+    for (auto neighbor : neighbors) {
+      if (std::find(ringSystemAtoms.begin(), ringSystemAtoms.end(),
+                    neighbor->getIdx()) == ringSystemAtoms.end()) {
+        neighborsOutsideTemplate++;
+      } else {
+        neighborsInsideTemplate.push_back(neighbor->getIdx());
+      }
+    }
+    // if the atom has more than one neighbor outside the template and exactly
+    // two matched by it, there's a chance it's a substitution point, so we
+    // should check if it's pointing inside a ring
+    if (neighborsOutsideTemplate >= 1 && neighborsInsideTemplate.size() == 2) {
+      auto coords = ringSystemAtomsCoords.at(index);
+      auto averageCoords = [ringSystemAtomsCoords](const RDKit::INT_VECT &vec) {
+        RDGeom::Point2D average(0, 0);
+        for (auto atom : vec) {
+          average += ringSystemAtomsCoords.at(atom);
+        }
+        average /= vec.size();
+        return average;
+      };
+      auto averageNeighborCoords = averageCoords(neighborsInsideTemplate);
+      auto substitutionV = (averageNeighborCoords - coords);
+      substitutionV.normalize();
+      substitutionV *= -0.4;
+      for (auto ring : fusedRings) {
+        if (pointInsideRing(coords + substitutionV, ringSystemAtomsCoords,
+                            ring)) {
+          result += neighborsOutsideTemplate;
+          break;
+        }
+      }
+    }
+  }
+  return result;
+}
+
+bool EmbeddedFrag::matchToTemplate(const RDKit::VECT_INT_VECT &fusedRings,
+                                   const RDKit::INT_VECT &ringSystemAtoms,
                                    unsigned int ring_count) {
   CoordinateTemplates &coordinate_templates =
       CoordinateTemplates::getRingSystemTemplates();
 
-  // only look for an exact match to the ring system because our method of
-  // completing rings from a template isn't reliably better than not using
-  // a template at all
```

Comment:
Isn't this partially true in the scope of this function? I think this would make sense to be removed in your other review